File: Arrays/Selection-sort.ts
const findSmallestPosition = <T>(arr: T[]): number => {
  let smallestIndex = 0;
  let smallestElement = arr[smallestIndex];

  for (let index: number = 1; index < arr.length; index++) {
    if (arr[index] < smallestElement) {
      smallestElement = arr[index];
      smallestIndex = index;
    }
  }

  return smallestIndex;
};

const selectionSort = <T>(arr: T[]): T[] => {
  let sortedArray: T[] = [];
  const arrayLength: number = arr.length;

  for (let index: number = 0; index < arrayLength; index++) {
    const smallestPosition = findSmallestPosition(arr);
    sortedArray.push(arr.splice(smallestPosition, 1)[0]);
  }

  return sortedArray;
};

const testArr = [5, 3, 6, 2, 10];

console.log("selectionSort", selectionSort(testArr));

File: Arrays/Selection-sort.py
def find_smallest(arr):
    smallestIndex = 0
    smallestValue = arr[smallestIndex]

    index = 1

    while index < len(arr):
        if arr[index] < smallestValue:
            smallestValue = arr[index]
            smallestIndex = index

        index += 1

    return smallestIndex


def selection_sort(arr):
    sorted_array = []
    arr_len = len(arr)

    for i in range(arr_len):
        smallest_value_position = find_smallest(arr)
        sorted_array.append(arr.pop(smallest_value_position))

    print(sorted_array)
    return sorted_array


testArr = [5, 3, 6, 2, 10]

selection_sort(testArr)
